### ISSUE TITLE
Parquet materialization 

### ISSUE THREAD

**Issue by jpmmcneill** [jwills/dbt-duckdb#19]
👋

Hey!

Something I've been thinking about is using parquet as a _materialization_. To be specific, one where `dbt-duckdb` would use the underlying duckdb connection as a shim to edit parquet rather than adding some tables into the `.duckdb` file.

I'm not sure if it's possible to override what `ref` does in an adapter, but this would roughly need to do two things:

1. Make a new materialization that just took the query for some model and executes
`copy (select ...) to 'model_name.parquet' (format parquet)` rather than`create table model_name as (select ...)` for parquet materialized models.
2. Update ref to use `read_parquet(ref'd model name)` rather than schema.table_name _if the model being referenced was materialized as parquet_.

Would love to hear what other people think / if this would be useful! 

This is loosely related to: https://github.com/jwills/dbt-duckdb/issues/15

**Comment by jpmmcneill** [comment]
It goes without saying that I'm happy to take a stab at it, but i'd love to get a temperature check from folks first 😄

**Comment by jwills** [comment]
Both of those things seem doable to me, but there are a few dragons we need to be mindful of. The big one is handling backups: when dbt materialized tables/views, it creates backups of the old relations via `rename` operations before it actually materializes the new one, so that if the new materialization fails for some reason, we haven't destroyed the old relation and can recover its contents. I just did some local testing of this idea with some `COPY (SELECT ...) TO 'output.parquet'` and it seemed like things mostly worked (i.e., various kinds of query failures did _not_ seem to blow away the file until the query succeeded), but I'd want to be sure to test this out pretty thoroughly before recommending it to folks as a more general pattern.

Second, it seems like a pretty easy lift to do this for CSV files at the same time, right? You would need a way for the `config(...)` macro block to accept options related to e.g. compression codecs for either format, and then of course [CSVs have a bunch more parameters](https://duckdb.org/docs/data/csv) you can config, but it seems like it's the same idea and would probably be generally useful as well.

Yeah, with those two things in mind, I think you could build something off of the changes in this branch to introduce a `DuckDBRelation`: https://github.com/jwills/dbt-duckdb/compare/master...jwills_file_based_dbs where you could be mindful of what your target materialization was and implement the `render` method accordingly.

Have you created custom dbt materialization strategies before? They aren't super-common in dbt projects, but we had to do a bunch for the Materialize adapter: https://github.com/MaterializeInc/materialize/tree/main/misc/dbt-materialize/dbt/include/materialize/macros/materializations

**Comment by jwills** [comment]
Ah there's also some work we would need to do to make the adapter aware of whether or not a relation existed already or not (I'm thinking of `list_relations_without_caching` as an example) which would involve some filesystem operations:

https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/adapters/base/impl.py#L178

Also would we refer to the parent directory of a set of parquet files as a database or a schema in this system?

**Comment by jpmmcneill** [comment]
Thanks @jwills. Agreed that this part of dbt has definite dragons!

I might also see what the good folks in slack think, as I'm not sure if this feature would be useful to people or if it's just a fever dream that should go away 😂

**Comment by jpmmcneill** [comment]
> Ah there's also some work we would need to do to make the adapter aware of whether or not a relation existed already or not (I'm thinking of `list_relations_without_caching` as an example) which would involve some filesystem operations:
> 
> https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/adapters/base/impl.py#L178
> 
> Also would we refer to the parent directory of a set of parquet files as a database or a schema in this system?

I'm really not too sure, to be honest I hadn't thought this far ahead 😂

It'd make sense to me for this to be more like a schema.

So for example, if I had my profile pointing to `quack/mallard.duckdb`, then materializing a table `blah` as a file would send it to

`quack/schema/blah.<extension>`

The only reason I see schema being useful here is that it's used a bit more than database during model config in my experience.

If a db was also defined I guess you could go to `quack/database/schema/blah`

Does that make sense?

**Comment by tomsej** [comment]
Hey @jpmmcneill I just created a new PR that could solve your problem - https://github.com/jwills/dbt-duckdb/issues/15#issuecomment-1291019714. Can you take a look if it satisfies your needs? Thank you!

**Comment by jpmmcneill** [comment]
@tomsej thanks for this!

I don't think it solves this issue exactly - but it's definitely helpful. I think this could coexist alongside the materialisation that I'm writing for this.

The reason I don't think these are the same is that https://github.com/jwills/dbt-duckdb/pull/33 feels like it's the same as running a post hook that unloads the file.

It feels quite heavy handed to both materialize the file in question into both the duckdb DB and the environment. The complication of only going via a file is that you need to handle the relation part properly as Josh mentioned above.

However, a materialization of "go into duckdb AND a file" feels like a nice feature, and this is pretty much what you've implemented!

**Comment by jpmmcneill** [comment]
I've made some progress on a branch. I'll push it up here and open a PR. There is one tricky part of the relation management that i'm still dealing with

**Comment by jpmmcneill** [comment]
it's over here:

https://github.com/jwills/dbt-duckdb/pull/34

Basically the issue that i'm seeing when testing this locally is that when the parquet / csv relation gets created and committed, the type that ref picks up (ie. relevant for the render method) isn't the one i'd expect.

**Comment by tomsej** [comment]
Hey @jpmmcneill thank you for kind words!  I am really looking forward to PR. I just want to make clear that data is not in `duckdb` and file. `DuckDB` works as a processing engine and metadata registry, there are no data in case of all materializations are `external_table`. The difference to post_hook is also in refering models; you can use data from file in other models and even in python models. My last goal was also to easily support `iceberg` and `delta` in the future.

Honestly, I am not really sure if there should be 2 ways for a similar thing, but it is on @jwills to decide. Can you elaborate more on your use case or problem you are trying to solve? If I get it, you do not want to explicitly store metadata of parquet files in the `DuckDB` but rather use directory structure to implicitly decide the database structure? Is it because you need multiple instances of `duckdb`? Or you want to use just as an engine so there should not be any data or metadata stored in DB? Sorry, I am just trying to understand if there is not a simple way...

Thank you!!!

**Comment by jpmmcneill** [comment]
> Hey @jpmmcneill thank you for kind words! I am really looking forward to PR. I just want to make clear that data is not in `duckdb` and file. `DuckDB` works as a processing engine and metadata registry, there are no data in case of all materializations are `external_table`. The difference to post_hook is also in refering models; you can use data from file in other models and even in python models. My last goal was also to easily support `iceberg` and `delta` in the future.
> 
> Honestly, I am not really sure if there should be 2 ways for a similar thing, but it is on @jwills to decide. Can you elaborate more on your use case or problem you are trying to solve? If I get it, you do not want to explicitly store metadata of parquet files in the `DuckDB` but rather use directory structure to implicitly decide the database structure? Is it because you need multiple instances of `duckdb`? Or you want to use just as an engine so there should not be any data or metadata stored in DB? Sorry, I am just trying to understand if there is not a simple way...
> 
> Thank you!!!

Thanks for the detail! If I read your pr correctly (which I definitely didn't previously - apologies 😅), if I materialise model x as a external_table as a CSV, dbt basically:
1. export the file as per the config 👏
2. create a view to that location (ie. `select * from x.csv` or similar 

The use case is basically using duckdb as an engine. In the case above, I don't think a `:memory:` type DB connection would properly work with suck external tables (ie. the view wouldn't exist and thus even if the file was already materialized the query wouldn't work as the db had just been spun up, and therefore there were no views created!

Hope that makes sense 😄

I think a combination of the two PRs might be helpful to give a fully lightweight (à la duckdb style) solution

**Comment by tomsej** [comment]
No worries and definitely there is absolutely no reason to apology. Your points are valid and it is truth that with `:memory:` type it would not work as expected. I have been already thinking about it and my idea was to add a new configuration parameter to `external_table`, let's call it `only_metadata` (sorry I am terrible at naming things :D). If this parameter is provided, e.g. something like `dbt run -s +model_name --vars '{"only_metadata": "true"}'` (or globally in `dbt_project.yaml`) it would spin up a new instance and for every `external_table` it would only create a view from file (basically just creating a metadata in the `:memory:` database). It should be really fast since it is just a metadata operation. Also, it would create only the necessary metadata due to `dbt` selectors, so the wrong data that do not relate would not break my query.

What do you think?

**Comment by matsonj** [comment]
Looking forward to seeing what you guys come up with - but I wanted to add some color here as I am handling this today in my [test project for dbt, duckdb, and parquet](https://github.com/matsonj/nba-monte-carlo).

I am using dbt-duckdb in two modes. First I am using it in "normal" mode, using duckdb database files. Then second, I am using ```:memory:``` database and dropping off parquet files with a post hook. I am using dbt ```target``` to manage my environment, although contemplating better ways to do this. So anyway, to drop tables off in parquet instead of duckdb, I simply do this: ```dbt run --target parquet```

What I landed on after some testing was the following dbt config block:
```
{{
    config(
        materialized = "view" if target.name == 'parquet' else "table",
        post_hook = "COPY (SELECT * FROM {{ this }} ) TO '/tmp/storage/{{ this.table }}.parquet' (FORMAT 'parquet', CODEC 'ZSTD');"
            if target.name == 'parquet' else " "
) }}
```

Then to reference it subsequently, I had to do this dumb trick:
```
FROM {{ "'/tmp/storage/upstream_model.parquet'" if target.name == 'parquet'
        else ref( 'upstream_model' ) }} E
```

Anyway - this *works* as is. I'm not saying that its good or bad, just that this is how I'm doing it.

**Comment by jpmmcneill** [comment]
Thanks for the context @matsonj!

I'm sure that there is something more refined that can be done here to handle the large replication of if / else blocks that this probably took!

Do you see the pattern of using a duckdb view / table for development and then parquet for production being a common one?

**Comment by jwills** [comment]
@matsonj I agree with you that the problem is transitory, but I think it's "transitory" in the same way that the Fed said inflation was "transitory"-- i.e., it will go away, but we should expect it to be with us for awhile, and so I think some sort of first-class support for parquet is warranted. I'm curious what an ideal parquet materialization config would look like for you in a project where you were expecting to materialize many (or maybe even all) of your dbt models as parquet files?

**Comment by jwills** [comment]
@jpmmcneill I also saw the issue where a `ref`'d `relation` object no longer has a `type` associated with it for some reason; I was going to bring that conversation into the #adapter-ecosystem channel on the dbt Slack to talk it over with Anders et al. My gut feel here is that for `ref`/`source` purposes, dbt doesn't generally need to know the distinction between e.g. a table and a view (since they look identical from a query perspective) and so it hasn't been an issue for them.

**Comment by jpmmcneill** [comment]
@jwills I think the issue with refs is related to the low level implementation of ref:

https://github.com/dbt-labs/dbt-core/blob/19167842873ea882c2dbb70aeb525953812d8ef3/core/dbt/context/providers.py#L463

which uses:

https://github.com/dbt-labs/dbt-core/blob/19167842873ea882c2dbb70aeb525953812d8ef3/core/dbt/adapters/base/relation.py#L243

& the following `create_from_node`. This doesn't look to be passing types correctly. As far as I can see, when a model gets materialized as type `foo` and a corresponding relation is returned, a new relation will be created via the ref.

**Comment by jwills** [comment]
@jpmmcneill ahh, excellent detective work-- let's alert the powers that be

**Comment by jpmmcneill** [comment]
Thanks @jwills - hopefully i'm on the right track as I'm not super familiar with the inner workings of dbt 😄

I'll post this into #adapter-ecosystem shortly & hopefully a guru can either confirm or point me in the right direction

**Comment by jwills** [comment]
@jpmmcneill I went right to #dbt-core-development, will cc you

**Comment by tomsej** [comment]
Sorry, I had a busy day today, so had no time to think too much about this or to implement my idea. I am not really sure if I copy the overall direction of this discussion. So I will try to sum up some general use-cases and outline a solution, feel free to add anything:

### Hybrid models

In this use-case the most of the models are materialized inside DuckDB just a bunch of selected models are materialized outside (as a parquet or csv file). So, it makes sense to keep the metadata inside of DB. Think this use-case will be the most common.

**Proposed solutions:**
- Think this use-case is fully supported by #33. It is also pretty close to what @matsonj proposed (that is actually pretty clever), it just adds some syntactic sugar to it. Like there is no need to create post_hooks with IF statements or having if statements inside ref, etc.

### DuckDB as an engine

In this use-case no data are actually stored inside DuckDB (so the `:memory:` option is used). All the models are materialized as files (csv, parquet, etc.) and DuckDB behaves as processing engine. Storing metadata inside DuckDB is therefore not possible since after the connection is closed all data are lost.

**Proposed solutions:**
- proposition from @jpmmcneill, that overload the render function of Relation object. This solution expects the folder structure to behave as an implicit metadata/database storage.
  - I really like the idea, but I am not convinced that replacing an `information_schema` with folder structure is a good idea. `dbt` is most of the time creating a full DAG of dependencies, so if any file (even not related to your dbt selector) is manually deleted the dbt will fail. I can imagine it cat be hard to debug and error-prone. Information schema also contains metadata for faster querying the data.
- I was thinking about this use-case, but somehow keeping the metadata inside `information_schema`. I stole this idea from https://github.com/dbt-labs/dbt-external-tables. The proposition from the #33 will additionally operate in the 2 modes:
   - **standard** - first step of the meterialization would be checking if `parquet` or `csv` exists in specified location. If yes, then it executes `create view as select * from <locacion>` (so basically it will create a metatada in the information schema. This operation should be really fast). If the file does not exist, then it will dump sql result into `parquet` and create a view on top.
   - **full-refresh** - it will always export to `parquet` and create a view.
   
There is also a special use-case from @matsonj that likes deciding if the models should be materialized inside DuckDB or in files by specifying the target. Think it can be solved with #33 too, with just one if statement `materialized = "external_table" if target.name == 'parquet' else "table"` on model or project level.

**Comment by tomsej** [comment]
I really like the comment from @matsonj. I once tried to write a `parquet` parser, and the specification was messy. Think DuckDB is a much better format. DuckDB is not only a database. It is also API, and it is becoming a wonderful storage format.

**Comment by matsonj** [comment]
I think the comment re: (```iceberg``` or ```delta```) is spot on. Implementing this materialization type will enable "true" multi-node dbt-duckdb because you can break your DAG apart and feed branches of your DAG to specific nodes. This allows horizontal scale really only seen in big data platforms but with BYO compute & storage.

**Comment by jpmmcneill** [comment]
Hey @tomsej, thanks for all the detail! A few small questions:

>   * I really like the idea, but I am not convinced that replacing an `information_schema` with folder structure is a good idea. `dbt` is most of the time creating a full DAG of dependencies, so if any file (even not related to your dbt selector) is manually deleted the dbt will fail. I can imagine it cat be hard to debug and error-prone. Information schema also contains metadata for faster querying the data.

Why would dbt fail? I might be missing something very obvious 😂

Also - given that we're dealing with parquet, is metadata in the information schema useful?

Unrelated to the two above, but I also think that we should consider developer experience a little. Personally, when I write a dbt model that fails in execution (which happens much more frequently than I'd like to admit) I quite often need to debug it by looking into the compiled sql and executing it locally to figure out the underlying problem. This is especially true for more complex queries.

I'm wondering if creating a view in a `memory` type connection will make the process to use the compiled code a good bit more difficult to copy / pasta somewhere else as the sql would query the view (which wouldn't exist outside of the connection) rather than the file directly.

**Comment by jwills** [comment]
@jpmmcneill many good questions, let me try to hit a couple of them here.

On why dbt would fail-- I think that the only _required_ feature from your PR that is missing is an impl of `list_relations_without_caching` on the adapter (which is usually implemented via queries to the `information_schema`), which I put into my attempt to make this work here: https://github.com/jwills/dbt-duckdb/compare/jwills_file_based_dbs?expand=1

I think it's worthwhile to compare your attempt and mine to illustrate what makes me uncomfortable going down the road I illustrate in the PR: we're essentially inventing our own database catalog system in `dbt-duckdb` in order to support this feature. It is admittedly a very simple database catalog (relations are parquet/csv files on disk, databases/schemas are directories-- although you and I differ on how we support schemas in our approaches), which is why I get why it's tempting to do it, but I think that going down this road is going to take us to a bad place where we're essentially reinventing a database catalog in python using a tool (dbt) that was never meant to do such a thing. The right place for the database catalog to live is in the database, which is why I like @tomsej's solution here w/views (an entry in the DuckDB catalog) backed by files on disk or object storage. If you think about this for a bit, and you still think I'm wrong, then I say let's go off and create a `dbt-parquet` project that uses the DuckDB engine but implements the database catalog purely as files on disk with whatever layout you like, without having the constraint of maintaining compatibility with whatever future features the DuckDB folks come up with. TBH I'm sure someone is going to do this at some point anyway, so it might as well be someone who I trust to do it well. ;-)

Second question: the metadata is required for the required `list_relations_without_caching` method and for `dbt docs` generation (column names/types.)

Third question, re: memory dbs: I agree with you, I don't think that a pure in-memory duckdb instance for a dbt run is a good idea. But to be clear we're not requiring anyone to run stuff in memory in order to use the parquet/csv materialization types, it's just something that a user can do if they are _extremely_ confident in their SQL abilities and the quality of their input data.

Re: in memory connections: I don't totally understand the issue you're raising, but in dbt-duckdb there is a single parent connection and all of the threads create their own child connection from it, which happens here: https://github.com/jwills/dbt-duckdb/blob/master/dbt/adapters/duckdb/connections.py#L109

**Comment by jpmmcneill** [comment]
Thanks @jwills. The above makes sense to me and I understand better (& agree with) the hesitancy to go down the pure file route now (questison 1). Thanks for clarifying, full steam ahead & I'm happy to help out with whatever 😄

in memory connections point is not relevant given the above 🚀

**Comment by matsonj** [comment]
Strongly agree with @jwills that it makes more sense to create its own adapter for ```dbt-parquet``` that leverages duckdb. That way there is a clear path for delta / iceberg that doesn't make ```dbt-duckdb``` super heavy. We want to keep this one very light for POC ideas IMO.

**Comment by tomsej** [comment]
I think @jwills wrote it better than I ever could. I am just giving my 2 cents. Yeah, parquet has a good metadata but e.g. csv do not. And we should support other formats too. In addition, the parquet metadata contains max/min statistics for every column, which is more useful for pruning and not necessary that useful for e.g. joining. Chances are that DuckDB is most likely counting some extra statistics for e.g. faster join statements, etc.
We, as a data community, already went the road of not using DBMS. It was in Hadoop before Hive Metastore was introduced. And it went messy quite soon. Ironically, reinventing DBMS is why Hannes created the DuckDB.

> Unrelated to the two above, but I also think that we should consider developer experience a little. Personally, when I write a dbt model that fails in execution (which happens much more frequently than I'd like to admit) I quite often need to debug it by looking into the compiled sql and executing it locally to figure out the underlying problem. This is especially true for more complex queries.

This is a good point. But this is already a case in dbt-core with incremental models. Think the easies solution would be to write compiled code into debug log. What do you think, @jpmmcneill?

**Comment by AlexanderVR** [comment]
> Why use parquet vs duckdb files? Imo the answer today is twofold: 1) parquet is more broadly compatible. 2) the duckdb file format is unstable.

I have third answer to @matsonj : duckdb files are a single homogenous file. Whereas Parquet files can be split up by table, schema, and partition, allowing efficient partial (re-)materializations when compute and storage are separated as is common with cloud workloads. This is exemplified by https://dagster.io/blog/duckdb-data-lake. 

Thought I would throw in my experience here as well.

@jpmmcneill 's idea, of driving everything through the filesystem is a tempting one. Confirming @jwills, I did try to implement such a `dbt-parquet` project. The main issue I ran into is that `parquet_scan(...)` is *not* a proper identifier. It is a table function. Under the hood `dbt` calls `relation.render()` in all kinds of situations, making this fiction hard to sustain outside of the basic "just compile this model please".

Registering each Parquet file as a view in duckdb, even if only in memory, makes for a nice separation between a relation's identifier and how we interact with it. So it was the natural evolution in what is now https://github.com/AlexanderVR/dbt-parquet

Now I had originally imagined the semantics of "parquet file as table" interfering with `dbt-duckdb`'s proper use of duckdb as a database, so I didn't seriously consider a PR to this repo. But I love @tomsej 's external materialization approach to keep these types distinct and all in the same system. A few questions + comments there:

1. how can Parquet files be used as sources?
2. can external tables have a default location based on schema, table name, and an additional `default_external_root` adapter parameter? Maybe an overrideable macro defines this location?
3. it's slick how external tables are created just with duckdb COPY commands. But at some point won't `dbt` need to manipulate the filesystem outside of duckdb? E.g. when external tables will lie in a folder that doesn't exist yet, when handling multi-part parquet files, or dropping/renaming tables and schemas like when testing with `dbt`.

**Comment by tomsej** [comment]
Thank you, @AlexanderVR those are great points. I apologize for a longer answer, I take it as a mental exercise on how I think about DuckDB. So please take it with the best intentions.

Regarding the Parquet file, I strongly agree. But I really think that what DuckDB is doing is a good thing. With current cloud workloads you have storage and API/engine separated, exactly as you wrote. With DuckDB **data is an API**. Imagine you have some interesting table in Snowflake and you want to share this data with the product team. You can either allow them to connect directly to Snowflake (which is expensive) or you will dump data to Parquet and create some kind of API on top. In every case, the data team is responsible for the API and data. With DuckDB you have a file, e.g. on S3, and a product team will incorporate DuckDB into their code. It means product is responsible for their API and data team only for the data.
With GDPR/CCPA the data is not only the asset any more, it is a liability too. Plus, analysts usually do not trust old data. What I want to say is that the amount of active data (by this I mean data that is actively processed/queried every day) is usually small (hundreds of GB), which is a good use-case for DuckDB file format. On the other hand, it is not good format for someday/maybe/archive data where parquet will always have e.g. better compression rate than DuckDB.

tldr; think both formats have their place in modern data stack. BTW currently working on baking Glue schema registrar baked into `dbt-duckdb` to make interoperation easier.

Regarding the splitting up DuckDB by table. I saw this PR https://github.com/duckdb/duckdb/pull/5048, that **perhaps** will make it possible, idk.

Now the main part, answering your questions:

> how can Parquet files be used as sources?

Think @jwills answered it somewhere else. Currently, you need to replace `identifier` with something like "read_parquet('fff.parquet')". In the future we should probably make it more user-friendly.

> can external tables have a default location based on schema, table name, and an additional default_external_root adapter parameter? Maybe an overrideable macro defines this location?

Good idea, I can take a look on that.

> it's slick how external tables are created just with duckdb COPY commands. But at some point won't dbt need to manipulate the filesystem outside of duckdb? E.g. when external tables will lie in a folder that doesn't exist yet, when handling multi-part parquet files, or dropping/renaming tables and schemas like when testing with dbt.

Good point. Honestly, I was aware of this, but I do not want to do it before it is **REALLY** necessary. It can get wild pretty soon, which is btw the reason why DuckDB is not doing that either. In addition, it is a case only if you are using local filesystem, with S3 it should work by default. My opinion is that if you are testing on localhost you should probably not use `external` materialization or you should handle the folder structure by yourself. But I appreciate any criticism of this opinion @AlexanderVR, @jwills.

**Comment by tomsej** [comment]
@AlexanderVR added the support for default external location in https://github.com/jwills/dbt-duckdb/pull/44/files.

**Comment by AlexanderVR** [comment]
@tomsej A few minor points:
- regarding https://github.com/jwills/dbt-duckdb/pull/44, doesn't the location need to be fully qualified? Not seeing where we specify the root directory here.
- I think abstractions like https://github.com/PyFilesystem/pyfilesystem2 can make the filesystem interactions less messy -- whenever you think it's worth tackling that challenge. If you're interested I can put up a draft PR illustrating this.
- I'm not that familiar with AWS Glue, but aren't there restrictions around nested arrays and structs?

Now, thanks for illustrating how you envision Duckdb providing an API over data. Let's leave aside the data "swamp"/archive use case for the rest of this discussion. 

I also deal normally in the range of 10s to 100s of GBs of "useful" active data so I think we have common goals here. It's just that the details around duckdb format vs parquet are making it hard for me to understand how to put the former into practice:

1. If the duckdb file is the point of contact between data team and the api, then managing the lifecycle of this "monolithic" asset I see involving a lot of excess data transfer and coordination. Especially given https://github.com/duckdb/duckdb/issues/77#issuecomment-1033224761. Even if you are fine with copying over the entire duckdb file each time it's updated, wouldn't we need to communicate with the API to have it reopen the duckdb connection to this new file?
2. Given atomic writes to S3 or local filesystems, a Parquet file per table approach allows finer-grained updates while avoiding the coordination problems. `duckdb`'s `parquet_scan()` appears to already have proper cache invalidation of parquet metadata which is nice :-)

You have probably thought much harder about this, so would appreciate any comments about how you envision (1).

**Comment by tomsej** [comment]
Thank you, @AlexanderVR! First work than play 😄:

> regarding https://github.com/jwills/dbt-duckdb/pull/44, doesn't the location need to be fully qualified? Not seeing where we specify the root directory here.

I did it a little bit differently to your proposition, but I think it is more flexible. Here are some examples:
- if you keep it as it is, it will store files in `<schema_name>/<table_identifier>.<format>`. You still can have specific location for certain models by specifying `location` in model's `config`.
- you can override a macro `external_location(format)` (create a new macro of the same name in `macro` folder) and e.g. add default prefix `s3://my-bucket/{{ this.schema }}/{{ this.identifier}}.{{ format }}`
- you can also override a macro and allow having root directory as project variable. So, the macro would look like, e.g. `{{ var('root_directory') }}/{{ this.schema }}/{{ this.identifier}}.{{ format }}`. Then, you can specify root folder in `dbt_project.yml` for whole project or on schema level.
- you can specify `location` in model config and use dbt variable - `location: var("root_directory") ~ "/my_awsome_table.parquet"`
....

Does this cover your requirements?

> I think abstractions like https://github.com/PyFilesystem/pyfilesystem2 can make the filesystem interactions less messy -- whenever you think it's worth tackling that challenge. If you're interested I can put up a draft PR illustrating this.

Interesting library. But I don't really think that we should add more libraries to `dbt-duckdb` than the bare minimum we currently have. It can always cause some dependency mismatch and so. When I was speaking about a mess, I was more pointing to this part `handling multi-part parquet files, or dropping/renaming tables and schemas` you wrote. If all you need is creating a folder if not exists, it should be easy even with default python libraries. Don't get me wrong, I am not saying that we should not do it, we will most likely implement it eventually, but we should know the use-case or be sure there is not some better way. Gosh, I already sound like a product manager. Or is it a produck manager 😄 ?

> I'm not that familiar with AWS Glue, but aren't there restrictions around nested arrays and structs?

Honestly, I do not remember. Will take a look on that. In my old code I had data type that looks something like `struct<type:string, accountid:string>`. But this feature will be optional and marked experimental, for sure.

**Comment by AlexanderVR** [comment]
Yes the macro should work just fine, thanks!

Definitely agree on caution around new libraries. I believe that makes you a pro-duck engineer not a manager.

Looking forward to the play.

**Comment by matsonj** [comment]
I agree that more libraries should not be added. As a practical note, I have been experimenting with connecting to s3 [in my MDS in a box project](https://github.com/matsonj/nba-monte-carlo/pull/29) and I think that it inadvertently takes the project in the wrong direction - mainly, it gets slower. 

The workflow implemented in the PR linked above both reads and writes from files located in S3 (admittedly, leveraging localstack, so its "fake S3" but i digress). What this introduces is unacceptable increase in run time. This one changes increases the amount of time to run my project from ~27s to ~2m30s. And these are not large files.

The pattern that should be implemented imo is to keep storage as close to duckdb as possible (i.e. inside your container) and then only at the end, write out to S3 (or another blog target). In that case, we should not even consider file management in scope except as already noted - dropping to an external location colocated with duckdb.

**Comment by AlexanderVR** [comment]
@matsonj where is your container running? I have heard second hand that S3 throughput can be just as fast as EBS on a decently-sized EC2 instance. However, if duckdb is mostly making tiny range requests that could cause the large slowdown.

**Comment by tomsej** [comment]
Pretty interesting @matsonj, performance tests are on my todo list. But this seems too much, like @AlexanderVR wrote. Can you try it on real AWS S3?

**Comment by jwills** [comment]
@tomsej two-cents on a couple of things:

First, I think the "write the files to some other directory or S3 bucket" case is common/easy enough that we should have a top-level config option like `external_root` that the default `external_location` macro uses when it's writing external tables, I'm happy to take a crack at it to demo what it looks like it if you don't want to or are otherwise occupied.

Second, on the sources stuff, I think the nicest solution would be to rip-off the macro construct from the `dbt-external-tables` repo, where we would config the sources with the `external:` and `location:` and `format:` tags and then do either a) a top-level `dbt run-operation stage_external_sources` to materialize them as views in the local duckdb file or b) hack the hell out of the `source()` macro construct to turn the rendered `source` into a filepath/duckdb function call instead of a view/table. I think this is a healthy chunk of work tho and independent enough of the other stuff we have going on that it can wait for 1.3.1, no need to block 1.3.0 on it.

**Comment by matsonj** [comment]
@tomsej @AlexanderVR yeah I can try on real S3, just have to do the setup stuff. In that case though, wouldn't it make the most sense to run the entire thing in AWS (instead of local container with S3 storage)? 

The testing I am seeing comes from the time to build in a GitHub action, so it is self contained. 

Let me know your thoughts here. I don't want to distract from the goal here but I think writing and reading intermediate transforms to/from S3 could be problematic for performance.

**Comment by tomsej** [comment]
@jwills 
Ad 1) Ohh yeah, this is even better! It makes a lot of sense. Just want to make sure, by top-level config you mean in `dbt_project.yaml`?

Ad 2) Think I am more keen to the second option. And agree that it should not block the 1.3.0 since we have your quick-fix.

**Comment by jwills** [comment]
re: 1, I meant on the `duckdb` profile, which is serialized via this class: https://github.com/jwills/dbt-duckdb/blob/master/dbt/adapters/duckdb/connections.py#L23 (i.e., where we specify settings/extensions/etc.) I think it's possible to tie it from there down into the adapter so the macro can access it.

On 2, sounds good to me, we will save it for 1.3.1, I'll put together an issue to describe the idea

**Comment by tomsej** [comment]
@matsonj thanks a lot, I am super interested! And nice project, btw!! I would like to try both, from localhost to S3 and from e.g. Fargate to S3. But I understand it is not that easy to spin it up.

Speaking of Github Action, think it is not the best tool for profiling. According to my experience, sometimes some runs are running too long for unknown reason.

> Let me know your thoughts here. I don't want to distract from the goal here but I think writing and reading intermediate transforms to/from S3 could be problematic for performance.

I mostly agree, also it is why I wrote about hybrid model. But on the other side; Databricks, EMR or even Snowflake external tables are using Parquet files and I did not experience that big difference between local filesystem and S3. It's also a question of what you compare it to, if just spinning up a new Spark cluster takes ~5min, than 2m30s is awesome :D

**Comment by matsonj** [comment]
@tomsej After some thought, I'm punting completely on localstack right now, I think it's more trouble than it's worth for the time being. I did however find a nice pattern for parquet handling (undoubtedly influenced by this conversation). I am using the duckdb ```:memory:``` database and then exporting files to parquet ```on-run-end``` with a macro that finds tables in a certain schema and copies them into the local file system. Way more detail can be found over on that project. Its so freaking fast, just amazing.

**Comment by jwills** [comment]
Friends, this was a truly epic thread and I'm grateful to everyone who played a part in making this happen. If there are no objections, I'm going to close this as Fixed in 1.3.1 and we can open more issues for bugs we find and new features that we would like to have. <3
